s3/pkg/service/object.go

```diff
@@ -262,12 +267,128 @@ func (s *s3Service) GetObject(ctx context.Context, in *pb.GetObjectInput, out *p
+//
+// See http://docs.aws.amazon.com/AmazonS3/latest/dev/Versioning.html
+func (s *s3Service) DeleteObject(ctx context.Context, in *pb.DeleteObjectInput, out *pb.DeleteObjectOutput) error {
 	log.Infoln("DeleteObject is called in s3 service.")
 
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	}()
+
+	bucket, err := s.MetaStorage.GetBucket(ctx, in.Bucket, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return nil
+	}
+
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed.")
+		return nil
+	}
+
+	// administrator can delete any resource
+	if isAdmin == false {
+		switch bucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if bucket.TenantId != tenantId && tenantId != "" {
+				log.Errorf("delete object failed: tenant[id=%s] has no access right.", tenantId)
+				err = ErrBucketAccessForbidden
+				return nil
+			}
+		} // TODO policy and fancy ACL
+	}
+
+	switch bucket.Versioning {
+	case utils.VersioningDisabled:
+		if in.VersioId != "" && in.VersioId != "null" {
+			err = ErrNoSuchVersion
+		} else {
+			err = s.removeObject(ctx, bucket, in.Key)
+		}
+	case utils.VersioningEabled:
+		// TODO: versioning
+		err = ErrInternalError
+	case utils.VersioningSuspended:
+		// TODO: versioning
+		err = ErrInternalError
+	default:
+		log.Errorf("versioing of bucket[%s] is invalid:%s\n", bucket.Name, bucket.Versioning)
+		err = ErrInternalError
+	}
+
+	if err == nil {
+		// TODO: enable cache
+		/*
+			yig.MetaStorage.Cache.Remove(redis.ObjectTable, obj.OBJECT_CACHE_PREFIX, bucketName+":"+objectName+":")
+			yig.DataCache.Remove(bucketName + ":" + objectName + ":")
+			yig.DataCache.Remove(bucketName + ":" + objectName + ":" + "null")
+			if version != "" {
+				yig.MetaStorage.Cache.Remove(redis.ObjectTable,
+					obj.OBJECT_CACHE_PREFIX, bucketName+":"+objectName+":"+version)
+				yig.DataCache.Remove(bucketName + ":" + objectName + ":" + version)
+			}*/
+	}
+
 	return nil
 }
 
+func (s *s3Service) removeObject(ctx context.Context, bucket *meta.Bucket, objectKey string) error {
+	log.Debugf("remove object[%s] from bucket[%s]\n", objectKey, bucket.Name)
+	obj, err := s.MetaStorage.GetObject(ctx, bucket.Name, objectKey, true)
+	if err == ErrNoSuchKey {
+		return nil
+	}
+	if err != nil {
+		log.Errorf("get object failed, err:%v\n", err)
+		return ErrInternalError
+	}
+
+	backendName := bucket.DefaultLocation
+	if obj.Location != "" {
+		backendName = obj.Location
+	}
+	backend, err := getBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend with err:", err)
+		return err
+	}
+	sd, err := driver.CreateStorageDriver(backend.Type, backend)
+	if err != nil {
+		log.Errorln("failed to create storage, err:", err)
+		return err
+	}
+
+	err = sd.Delete(ctx, &pb.DeleteObjectInput{Bucket: bucket.Name, Key: objectKey, VersioId: obj.VersionId})
```

Comment:
it is correct to delete meta first and then delete data in backend, or if we first delete data in backend successfully and then fail to delete meta, user can query object successfully but he cannot get that object.

Reply:
Thank you. I changed it to be like this: Mark the object as deleted in database, remove data in the backend, and then delete object from database.